### PR TITLE
docs(provider/kubernetes): Add a guide for Patch(Manifest) stage.

### PR DESCRIPTION
**Pull request by dibyom**
New PR since I accidentally closed #850 by running a `git push -d` instead of a `git push -f` :disappointed:  

Addressed all the comments and added a entry in the pipeline reference.

@dorbin @lwander

### CODE REVIEW

**Review by lwander — APPROVED** [review]
Was wondering why it was closed... LGTM again

File: reference/pipeline/stages/index.md [hunk 1]
```diff
@@ -169,6 +169,9 @@ Deploy a Kubernetes manifest YAML/JSON file.
 ### Find Artifacts From Resource
 Find artifacts from a Kubernetes resource.
 
+### Patch (Manifest)
+Patch an existing Kubernetes resource in-place.
```
Comment (avh121) [inline reference/pipeline/stages/index.md]: Would you mind expanding this description a little bit? Ideally, talking about the use cases behind the feature, and any important info about how this stage works. (See the bake stage description for a model of the sort of things that are good to include here -- basically explaining why someone would use this stage, and what they should be aware of when they do so).

The reason that some of these descriptions are only a sentence is that this doc is still a work in progress, and I decided it was more useful to get a skeleton out there than to delay it until it was 100% complete. I'd like to fully explain new stages that are added, though.

Thanks!

**Review by avh121 — CHANGES_REQUESTED** [review]

File: reference/pipeline/stages/index.md (hunk 1 above)
Comment (dibyom) [reply reference/pipeline/stages/index.md]: Done.

File: reference/pipeline/stages/index.md [hunk 2]
```diff
@@ -169,6 +169,11 @@ Deploy a Kubernetes manifest YAML/JSON file.
 ### Find Artifacts From Resource
 Find artifacts from a Kubernetes resource.
 
+### Patch (Manifest)
+Update an already existing Kubernetes resource in place using the [Kubernetes patch operation](https://kubernetes.io/docs/tasks/run-application/update-api-object-kubectl-patch/). Spinnaker can update the resource without knowing the entire resource manifest i.e. you have to specify only the portion of the manifest you want to update. For instance, the patch stage can be used to add a label or update a sidecar container image for a set of resources. It can also be used to implement a red-black or "rainbow" deployment strategy for Kubernetes by first deploying a new ReplicaSet and then patching the fronting service's selectors to point to the new ReplicaSet.
```
Comment (dibyom) [inline reference/pipeline/stages/index.md]: I'm writing a more in-depth guide on rainbow deployments. Will link to it here when complete.

File: guides/user/kubernetes-v2/patch-manifest/index.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+* [(Optional) Override artifacts in the patch content](#override-artifacts)
+
+  Optionally, you can override some artifacts (as fields) at run time (for
+    example, which Docker image to use.)
+
+* [(Optional) Override patch specific options](#specify-patch-options)
+
+## Specify the resource to patch
+
+Start by selecting the __Patch (Manifest)__ stage from the stage selector:
```
Comment (dorbin) [inline guides/user/kubernetes-v2/patch-manifest/index.md]: I'd number these.

1. Select the __Patch (Manifest)__ stage from the stage selector:

File: guides/user/kubernetes-v2/patch-manifest/index.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+{%
+  include
+  figure
+  image_path="./patch-manifest.png"
+%}
+
+To identify the Kubernetes resource to patch, specify the following required fields:
```
Comment (dorbin) [inline guides/user/kubernetes-v2/patch-manifest/index.md]: 2. Identify the Kubernetes resource to patch:

* __Account__
* etc.

File: guides/user/kubernetes-v2/patch-manifest/index.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+  The Kubernetes namespace that your resource is located in.
+
+* __Kind__
+
+  The Kubernetes Kind of your resource e.g. deployment, service etc.
+
+* __Name__
+
+  The name of your Kubernetes resource.
```
Comment (dorbin) [inline guides/user/kubernetes-v2/patch-manifest/index.md]: You can delete the periods from this line and the line under __Namespace__.

File: guides/user/kubernetes-v2/patch-manifest/index.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  image_path="./in-pipeline.png"
+  caption="Notice that by selecting __Text__ as the __Manifest Source__, we get
+  to enter the manifest YAML by hand."
+%}
+
+### Specify patch content dynamically
+
+Like the [Deploy (Manifest) stage](/guides/user/kubernetes-v2/deploy-manifest#specify-manifests-dynamically), you can also reference an artifact as the source if you are storing your patch content externally. The artifact must be a text file containing the patch content.
+
+You can also setup the pipeline to trigger based on changes to the patch content:
```
Comment (dorbin) [inline guides/user/kubernetes-v2/patch-manifest/index.md]: s/setup/set up/

File: guides/user/kubernetes-v2/patch-manifest/index.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+* [Consuming GitHub Artifacts](/guides/user/triggers/github)
+* [Consuming GCS Artifacts](/guides/user/triggers/gcs)
+
+Assuming you have declared an expected artifact upstream to your Patch
+manifest stage, you can reference it in the Patch configuration:
```
Comment (dorbin) [inline guides/user/kubernetes-v2/patch-manifest/index.md]: "Patch (Manifest)" right?

File: guides/user/kubernetes-v2/patch-manifest/index.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  get to pick which upstream artifact to deploy."
+%}
+
+> __☞ Note__: Make sure that the __Artifact Account__ field matches an account
+> with permission to download the manifest.
+
+
+## Override artifacts
+
+The patch stage also allows you to [override artifacts](/guides/user/kubernetes-v2/deploy-manifest#override-artifacts) like in the deploy manifest stage.
```
Comment (dorbin) [inline guides/user/kubernetes-v2/patch-manifest/index.md]: "Patch (Manifest)" again, right?

File: guides/user/kubernetes-v2/patch-manifest/index.md [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+For instance, say you have a pipeline with a patch manifest stage with the following patch content:
```
Comment (dorbin) [inline guides/user/kubernetes-v2/patch-manifest/index.md]: and here too

File: reference/pipeline/stages/index.md (hunk 2 above)
Comment (dorbin) [inline reference/pipeline/stages/index.md]: s/entire resource manifest i.e. you have to/entire resource (that is, you have to/

Plus a closing ) of course

Comment (dorbin) [inline reference/pipeline/stages/index.md]: I don't know about the rainbow strategy. I can't tell if you're presenting it as an alternate term for red/black (which, I hope not; it's already a synonym for blue/green I think!), or if it's just another strategy. If it is different, I would...
s/implement a red-black or "rainbow" deployment strategy/implement a red/black deployment strategy or "rainbow" deployment strategy/

(Note that it's red/black not red-black.)

**Review by dorbin — CHANGES_REQUESTED** [review]
Dibyo, just a few suggestions/comments/questions

File: reference/pipeline/stages/index.md (hunk 2 above)
Comment (avh121) [reply reference/pipeline/stages/index.md]: Unrelated nit: would you mind splitting up this paragraph into multiple lines, at the 80-100 character mark? (The way other long blocks of text are formatted in this doc). It makes it easier to review specific lines.

Comment (dibyom) [reply reference/pipeline/stages/index.md]: It is basically the same as red/black or blue/green. So, I've removed rainbow and kept just red/black to keep things consistent.

Comment (lwander) [reply reference/pipeline/stages/index.md]: It's not entirely the same, and this approach has specific consequences that make it different. For example, in the prior red/black strategy one could have one load balance manage multiple replica sets during normal deployments. Using this strategy, at most one replica set can be managed.

Comment (dibyom) [reply reference/pipeline/stages/index.md]: Split into two paragraphs.

Also, removed red/black and added "rainbow" back in with a link to an external article. Will replace the link with the guide I'm writing on it when complete.

Comment (dorbin) [reply reference/pipeline/stages/index.md]: Looks good now!

**Review by dorbin — APPROVED** [review]

File: reference/pipeline/stages/index.md [hunk 10]
```diff
@@ -169,6 +169,13 @@ Deploy a Kubernetes manifest YAML/JSON file.
 ### Find Artifacts From Resource
 Find artifacts from a Kubernetes resource.
 
+### Patch (Manifest)
+Update an already existing Kubernetes resource in place using the [Kubernetes patch operation](https://kubernetes.io/docs/tasks/run-application/update-api-object-kubectl-patch/). Spinnaker can update the resource without knowing the entire resource (that is, you have to specify only the portion of the manifest you want to update).
```
Comment (avh121) [inline reference/pipeline/stages/index.md]: nit: would you put line breaks here at the 80-100 character mark the way we do in the rest of the file?

**Review by avh121 — APPROVED** [review]

**Comment by dibyom** [comment]
Can someone with push access merge this? Thanks!

**Pull request merged**